fedimint-client/src/lib.rs

```diff
@@ -1,6 +1,425 @@
+    pub DynGlobalClientContext(Arc<IGlobalClientContext>)
+}
+
+impl DynGlobalClientContext {
+    /// Waits till consensus has been achieved on the transaction and it was
+    /// accepted by consensus.
+    pub async fn await_tx_accepted(
+        &self,
+        operation_id: OperationId,
+        txid: TransactionId,
+    ) -> InactiveState {
+        let state = OperationState {
+            operation_id,
+            state: TxSubmissionStates::Accepted { txid },
+        };
+        self.await_inactive_state(state.into_dyn(TRANSACTION_SUBMISSION_MODULE_INSTANCE))
+            .await
+    }
+
+    /// Waits till the transaction is either rejected on submission or after
+    /// consensus has been achieved on it.
+    pub async fn await_tx_rejected(
+        &self,
+        operation_id: OperationId,
+        txid: TransactionId,
+    ) -> InactiveState {
+        let state = OperationState {
+            operation_id,
+            state: TxSubmissionStates::Rejected { txid },
+        };
+        self.await_inactive_state(state.into_dyn(TRANSACTION_SUBMISSION_MODULE_INSTANCE))
+            .await
+    }
+}
+
+impl<T> From<Arc<T>> for DynGlobalClientContext
+where
+    T: IGlobalClientContext,
+{
+    fn from(inner: Arc<T>) -> Self {
+        DynGlobalClientContext(inner)
+    }
+}
+
+impl GlobalContext for DynGlobalClientContext {}
+
+// TODO: impl `Debug` for `Client` and derive here
+impl Debug for ClientInner {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        write!(f, "ClientInner")
+    }
+}
+
+#[apply(async_trait_maybe_send!)]
+impl IGlobalClientContext for ClientInner {
+    fn api(&self) -> &(dyn IFederationApi + 'static) {
+        self.api.as_ref()
+    }
+
+    async fn finalize_and_submit_transaction(
+        &self,
+        dbtx: &mut DatabaseTransaction<'_>,
+        operation_id: OperationId,
+        tx_builder: TransactionBuilder,
+    ) -> anyhow::Result<TransactionId> {
+        ClientInner::finalize_and_submit_transaction(self, dbtx, operation_id, tx_builder).await
+    }
+
+    async fn await_active_state(&self, state: DynState<DynGlobalClientContext>) -> ActiveState {
+        ClientInner::await_active_state(self, state).await
+    }
+
+    async fn await_inactive_state(&self, state: DynState<DynGlobalClientContext>) -> InactiveState {
+        ClientInner::await_inactive_state(self, state).await
+    }
+}
+
+pub struct Client {
+    inner: Arc<ClientInner>,
+}
+
+impl Client {
+    pub fn context(&self) -> DynGlobalClientContext {
+        DynGlobalClientContext::from(self.inner.clone())
+    }
+
+    /// Add funding and/or change to the transaction builder as needed, finalize
+    /// the transaction and submit it to the federation.
+    pub async fn finalize_and_submit_transaction(
+        &self,
+        operation_id: OperationId,
+        tx_builder: TransactionBuilder,
+    ) -> anyhow::Result<TransactionId> {
+        let mut dbtx = self.inner.db.begin_transaction().await;
```

Comment:
Is this a readonly database transaction?

Reply:
no, I just forgot to commit and was about to spend some quality time with my debugger :stuck_out_tongue_winking_eye: thx for the warning ^^